rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -96,6 +98,274 @@ object SpatialMatcher {
 
   def emptyMap: FreeMap = Map.empty[Int, Par]
 
+  case class ParCount(val sends: Int = 0,
+                      val receives: Int = 0,
+                      val evals: Int = 0,
+                      val news: Int = 0,
+                      val exprs: Int = 0,
+                      val matches: Int = 0,
+                      val ids: Int = 0,
+                      val bundles: Int = 0) {
+    def binOp(op: (Int, Int) => Int, other: ParCount): ParCount =
+      ParCount(
+        sends = op(sends, other.sends),
+        receives = op(receives, other.receives),
+        evals = op(evals, other.evals),
+        news = op(news, other.news),
+        exprs = op(exprs, other.exprs),
+        matches = op(matches, other.matches),
+        ids = op(ids, other.ids),
+        bundles = op(bundles, other.bundles)
+      )
+    def min(other: ParCount): ParCount = binOp(math.min, other)
+    def max(other: ParCount): ParCount = binOp(math.max, other)
+    def +(other: ParCount): ParCount   = binOp(saturatingAdd, other)
+    // Only saturates going from positive to negative
+    def saturatingAdd(l: Int, r: Int): Int = {
+      val res = l + r
+      (res | -(if (res < l) 1 else 0)) & ~Int.MinValue
```

Comment:
They aren't BigInt's so, yes, you can have res < l due to wraparound
We're clearing the sign bit, so it makes more sense to use ~Int.MinValue

Reply:
No, in some cases, we're passing in the max value, because these are bounds. We need to not wrap in that case.

As to the second case, yes. It's a branch free version of that.

Reply:
1. Right, thanks.
2. OK, tbh I had to refresh binary operations to understand that one liner.